Notification based test watchdog

jenswi-linaro opened the pull request:
This introduces a test watchdog based on notifications. The primary purpose is to later be able to exercise asynchronous notifications with ns-virtualization, but to I'm trying a smaller scope to help the review.

The test watchdog is based on the introduced callout service which handles calling a callback at a certain timeout or period. The callout service is currently driven by the secure generic timer, but can easily be driven by another timer if needed.

jenswi-linaro commented:
It took a while, but CI is happy with the PR now.

etienne-lms reviewed (commented):
Comments for commit "core: add callout service".
Out of curiosity, why did you choose _callout_ naming for this feature. It is about delayed calls, right?

etienne-lms commented on core/arch/arm/kernel/generic_timer.c:
```diff
@@ -0,0 +1,65 @@
+{
+	uint64_t freq = read_cntfrq();
+
+	return (freq * timeout_ms) / 1000;
```
could use `arm_cnt_us2cnt()` from _core/arch/arm/kernel/delay_arm.h_

etienne-lms commented on core/arch/arm/plat-vexpress/main.c:
```diff
@@ -255,3 +256,13 @@ int psci_cpu_on(uint32_t core_id, uint32_t entry, uint32_t context_id)
+
+	return TEE_SUCCESS;
+}
+nex_early_init(init_callout_service);
```
add an empty line above.

etienne-lms reviewed (commented):
For commit "core: tests: add a notification test watchdog": the commit message is not very verbose about the test, what's expected etc.. Could you add few details?

jenswi-linaro commented:
> Comments for commit "core: add callout service". Out of curiosity, why did you choose _callout_ naming for this feature. It is about delayed calls, right?

It's something I hacked many years ago, initially inspired by https://github.com/freebsd/freebsd-src/blob/main/sys/sys/callout.h but rewritten so much that only the name remained. If you can think of a better **and** shorter name I don't mind renaming it.

jenswi-linaro commented on core/include/kernel/callout.h:
```diff
@@ -0,0 +1,112 @@
+ * @callback is called from an interrupt handler so thread resources must
+ * not be used. The main callout service lock is held while @callback is
+ * called so callout_rem() and callout_add() can't be used, but it is safe
+ * to modify @ms if the call period should be changed. @callback returns
```
Makes sense, that's a nice simplification.

jenswi-linaro commented on core/include/kernel/callout.h:
```diff
@@ -0,0 +1,112 @@
+ * @ms:		time to next callout in milliseconds
+ *
+ * Adds a callout to the callout service with an associated callback
+ * function @callout should be assigned and will be called in @ms
```
I'll update, thanks.

jenswi-linaro commented on core/include/kernel/callout.h:
```diff
@@ -0,0 +1,112 @@
+ *
+ * The callout structure can reside in global data or on the heap. It's
+ * safe to embed it inside another struct, but it must not be freed until
+ * removed again with callout_rem() or equivalent.
```
Yes

jenswi-linaro commented on core/include/kernel/callout.h:
```diff
@@ -0,0 +1,112 @@
+ *
+ * The callout service is initialized with the supplied timer interface
+ */
+void callout_service_init(const struct callout_timer_desc *desc);
```
Sure. Would you mind if that's something we add at a later stage?

etienne-lms replied:
I'm fine it's considered in a later change.

jenswi-linaro commented on core/kernel/callout.c:
```diff
@@ -0,0 +1,168 @@
+	TAILQ_INSERT_TAIL(&callout_head, co, link);
+}
+
+static void schedule_next_timeout(const struct callout_timer_desc *desc)
```
OK

jenswi-linaro commented on core/kernel/callout.c:
```diff
@@ -0,0 +1,168 @@
+	 * schedule_next_timeout() saves the core it was last called on.
+	 * If if there's a mismatch here it means that another core has be
+	 * scheduled for the next callout, so there's no work to be done
+	 * for this core.
```
Good point, I'll update.

jenswi-linaro commented on core/include/kernel/callout.h:
```diff
@@ -0,0 +1,112 @@
+/*
+ * struct callout - callout reference
+ * @callback:	function to be called when a callout expires, internal use only
+ * @ctrval:	callout expiry time counter value, internal use only
```
Using milliseconds has the drawback that we may need to divide the time counter value with the frequency quite often. I suppose it would mean to implement `timer_now()` as
`return barrier_read_counter_timer() * 1000 / read_cntfrq()`.

Will `expiry_value` work better?

etienne-lms commented:
shorter than 6 chars, you're quite challenging :)
`delayed_call`, `periodic_call`.
Well, I guess `callout` is ok since `struct callout` inline description comment well describes the scope and purpose of the feature.

jenswi-linaro commented:
Force push to fix the checkpatch issue.

jforissier reviewed (commented):
- For "core: arm64: add {read,write}_cntps_cval()":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "core: add callout service":
  s/will be converted to in/will be converted to ticks in/
  s/ If if/If/
  s/has be scheduled/has been scheduled/
  I am not sure about the name of `is_per_cpu`. If the underlying timer ensures that the interrupt is always received by the same core, then `is_per_cpu` doesn't matter, does it? (and it should preferably set to `false` to avoid some overhead). If the timer interrupt picks one (and only one) random core, then `is_per_cpu` needs to be `true`. This sounds backwards to me.
- For "core: define generic callout service initializer":
  s/shall provide and/shall provide an/ (in commit description)
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "plat-vexpress: initialize callout service":
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "plat-vexpress: qemu_armv8: define IT_SEC_PHY_TIMER":
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "core: tests: add a notification test watchdog":
  s/uart/UART/ in the commit description
  The trace "D/TC:0   notif_send_async:93 0x0" every second doesn't seem to stop, that's a bit annoying.

jenswi-linaro commented on core/arch/arm/kernel/sub.mk:
```diff
@@ -2,6 +2,7 @@ srcs-y += rpc_io_i2c.c
 srcs-y += idle.c
 
 srcs-$(CFG_SECURE_TIME_SOURCE_CNTPCT) += tee_time_arm_cntpct.c
+srcs-$(CFG_ARM64_core) += generic_timer.c
```
With minor tweaking, it should be usable in AArch32 also. But I can rename it until that is done if you like.

jenswi-linaro commented:
Force push to apply tags and fix up "plat-vexpress: initialize callout service"

jenswi-linaro commented:
> I am not sure about the name of `is_per_cpu`. If the underlying timer ensures that the interrupt is always received by the same core, then `is_per_cpu` doesn't matter, does it? (and it should preferably set to `false` to avoid some overhead). If the timer interrupt picks one (and only one) random core, then `is_per_cpu` needs to be `true`. This sounds backwards to me.

`is_per_cpu` is a description of the implementation. Timers can be global or per CPU. If it's per CPU it will only be manipulated by the current CPU. That means that if a new time is set on a different CPU we'll have an updated timer running on another CPU. Instead of updating that timer, we're just ignoring the interrupt from it.

jenswi-linaro commented:
Added "core: notif_send_async(): remove debug print" to remove the annoying debug print.

jforissier commented:
> > I am not sure about the name of `is_per_cpu`. If the underlying timer ensures that the interrupt is always received by the same core, then `is_per_cpu` doesn't matter, does it? (and it should preferably set to `false` to avoid some overhead). If the timer interrupt picks one (and only one) random core, then `is_per_cpu` needs to be `true`. This sounds backwards to me.
> 
> `is_per_cpu` is a description of the implementation. Timers can be global or per CPU. If it's per CPU it will only be manipulated by the current CPU. That means that if a new time is set on a different CPU we'll have an updated timer running on another CPU. Instead of updating that timer, we're just ignoring the interrupt from it.

I fail to imagine the use case. Can you add a test where `is_per_cpu` needs to be `true`? Otherwise I'd rather introduce it later when there is a real use case.

jforissier commented on core/tests/notif_test_wd.c:
```diff
@@ -0,0 +1,126 @@
+	wd->pending = true;
+	notif_send_async(NOTIF_VALUE_DO_BOTTOM_HALF);
+
+	return true;
```
Perhaps the test should stop re-scheduling itself at some point, when one can see that everything works as expected?

jenswi-linaro replied:
It's interesting to see that xtest can be run without any timeouts. I have coming patches where this works with FF-A and virtualization where two guests run this in parallel. I assume a debug print once a minute (per guest) isn't too much logging after the initial 10.

jenswi-linaro commented:
> I fail to imagine the use case. Can you add a test where `is_per_cpu` needs to be `true`? Otherwise I'd rather introduce it later when there is a real use case.

I'm sorry, I forgot to update "core: arm64: implement timer_init_callout_service()" when I added this `is_per_cpu` during the review. Perhaps it's more clear now.

jforissier commented:
- For "core: tests: add a notification test watchdog":
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>` 
- For "core: notif_send_async(): remove debug print":
  `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
Squashed and rebased on master.
These two commits have not been fully reviewed yet:
- core: add callout service
- core: arm64: implement timer_init_callout_service()

jforissier commented on core/include/kernel/callout.h:
```diff
@@ -0,0 +1,127 @@
+ *			unit
+ * @get_now:		get the current counter value
+ * @is_per_cpu:		flag to indicate if this timer is per cpu (true) or
+ *			global (false).
```
I wouldn't mind more details here. How about: "When true, callout_add() saves the current core number. When an interrupt is triggered, the callout service will check the core number again and only invoke the callback if the core is the same."?

jforissier reviewed (commented):
For "core: add callout service": one comment below. In any case:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented on core/arch/arm/plat-vexpress/main.c:
```diff
@@ -256,3 +257,14 @@ int psci_cpu_on(uint32_t core_id, uint32_t entry, uint32_t context_id)
+#if defined(IT_SEC_PHY_TIMER) && !defined(CFG_CORE_SEL2_SPMC)
+static TEE_Result init_callout_service(void)
+{
+	timer_init_callout_service(interrupt_get_main_chip(), IT_SEC_PHY_TIMER);
```
Still wondering about `is_per_cpu` here :wink: what happens when `IT_SEC_PHY_TIMER` is triggered? Is it delivered to only one core (and in this case, always the same or not?), to all cores, to a subset of all cores...?

jenswi-linaro replied:
There is one timer per CPU, and each of them only triggers a PPI on that CPU. The PPI is triggered only on the CPU where the generic timer expires.

jforissier reviewed (commented):
- For "core: arm64: implement timer_init_callout_service()":
  `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

jenswi-linaro commented:
I've added more description to `struct callout_timer_desc`, I hope this helps.

jforissier commented:
> I've added more description to `struct callout_timer_desc`, I hope this helps.

Fine with me. Perhaps with `@` in front of `set_next_timeout()` since it is a member of the structure.

jenswi-linaro commented:
Squashed and tags applied. Should be ready to merge now, unless @etienne-lms or someone else has something to add.

etienne-lms commented on core/tests/notif_test_wd.c:
```diff
@@ -0,0 +1,126 @@
+{
+	struct periodic_data *d = container_of(co, struct periodic_data,
+					       callout);
+	TEE_Time t = { };
```
`__maybe_unused`

jenswi-linaro replied:
This is always used by `tee_time_get_sys_time()`, even if we may throw away the result if debug prints are disabled. So I'm not sure that `__maybe_unused` is accurate.

etienne-lms commented on core/tests/notif_test_wd.c:
```diff
@@ -0,0 +1,126 @@
+
+	if (wd->pending)
+		wd->timeout_count++;
+	wd->call_count++;
```
Here we expect `wd->timeout_count == wd->call_count`, right?
Maybe an test PTA could report to a new xtest test ID whether this is true or not, so that regression test can check the notif and callout frameworks behave as expected.

jenswi-linaro replied:
No, we expect `wd->timeout_count` to be `0`, but the PTA thing sounds like a nice extension.

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` with few minor comments to address.

jenswi-linaro commented:
Addressed comments, squashed in the small fixes, and applied the tag.

jforissier commented:
This PR breaks `make check CFG_WITH_PAGER=y` on QEMUv8. See https://github.com/OP-TEE/optee_os/pull/6747.
